db/multi.go

```diff
@@ -50,39 +66,50 @@ func (mi *MultiInserter) Add(row []interface{}) error {
 // for gorp to use in place of the query's question marks. Currently only
 // used by .Insert(), below.
 func (mi *MultiInserter) query() (string, []interface{}) {
-	questionsRow := strings.TrimRight(strings.Repeat("?,", mi.numFields), ",")
+	questionsRow := strings.TrimRight(strings.Repeat("?,", len(mi.fields)), ",")
 
 	var questionsBuf strings.Builder
 	var queryArgs []interface{}
 	for _, row := range mi.values {
+		// Safety: We are interpolating a string that will be used in a SQL
+		// query, but we constructed that string in this function and know it
+		// consists only of question marks joined with commas.
 		fmt.Fprintf(&questionsBuf, "(%s),", questionsRow)
 		queryArgs = append(queryArgs, row...)
 	}
 
 	questions := strings.TrimRight(questionsBuf.String(), ",")
 
+	// Safety: we are interpolating `mi.returningColumn` into an SQL query. We
+	// know it is a valid unquoted identifier in MariaDB because we verified
+	// that in the constructor.
 	returning := ""
-	if mi.retCol != "" {
-		returning = fmt.Sprintf(" RETURNING %s", mi.retCol)
+	if mi.returningColumn != "" {
+		returning = fmt.Sprintf(" RETURNING %s", mi.returningColumn)
 	}
-	query := fmt.Sprintf("INSERT INTO %s (%s) VALUES %s%s;", mi.table, mi.fields, questions, returning)
+	// Safety: we are interpolating `mi.table` and `mi.fields` into an SQL
+	// query. We know they contain, respectively, a valid unquoted identifier
+	// and a slice of valid unquoted identifiers because we verified that in
+	// the constructor. We know the query overall has valid syntax because we
+	// generate it entirely within this function.
+	query := fmt.Sprintf("INSERT INTO %s (%s) VALUES %s%s", mi.table, strings.Join(mi.fields, ","), questions, returning)
 
 	return query, queryArgs
 }
 
-// Insert performs the action represented by .query() on the provided database,
-// which is assumed to already have a context attached. If a non-empty retCol
-// was provided, then it returns the list of values from that column returned
-// by the query.
-func (mi *MultiInserter) Insert(exec Executor) ([]int64, error) {
+// Insert inserts all the collected rows into the database represented by
+// `queryer`. `queryer` is assumed to already have a context attached. If a
+// non-empty returningColumn was provided, then it returns the list of values
+// from that column returned by the query.
+func (mi *MultiInserter) Insert(queryer Queryer) ([]int64, error) {
 	query, queryArgs := mi.query()
-	rows, err := exec.Query(query, queryArgs...)
+	rows, err := queryer.Query(query, queryArgs...)
 	if err != nil {
 		return nil, err
 	}
 
 	ids := make([]int64, 0, len(mi.values))
-	if mi.retCol != "" {
+	if mi.returningColumn != "" {
 		for rows.Next() {
```

Comment:
The `ok` value cannot be an error, it's a boolean and can only take the values `true` or `false`. There are two ways it can become `false`: when there are no more rows, or when trying to get the next row resulted in some sort of error. When the first of those happens, `rows.Err()` remains `nil`. When the second of those happens, `rows.Err()` returns whatever error occurred.

In addition, `rows.Close()` returns the same error that `rows.Err()` would have returned, if there is such an error set on the `rows.lasterror` field. So when we check the error return from `rows.Close()`, we're checking that.